Use `keepdims` and maintain backward compatible for `keep_dims`

yongtang opened the pull request:
This fix tries to address the issue raised in #6815 where
both `keepdims` and `keep_dims` were used with inconsistency.

This fix changes related api to `keepdims` while at the same
time maintain backward compatible for `keep_dims` so that
use will not be impacted.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Use `keepdims` and maintain backward compatible for `keep_dims`

This fix tries to address the issue raised in 6815 where
both `keepdims` and `keep_dims` were used with inconsistency.

This fix changes related api to `keepdims` while at the same
time maintain backward compatible for `keep_dims` so that
use will not be impacted.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/python/ops/linalg_ops.py:
```diff
@@ -422,7 +422,7 @@ def svd(tensor, full_matrices=False, compute_uv=True, name=None):
 
 # pylint: disable=redefined-builtin
-def norm(tensor, ord='euclidean', axis=None, keep_dims=False, name=None):
+def norm(tensor, ord='euclidean', axis=None, keepdims=None, name=None, keep_dims=None):
```
please use @deprecated_args

yongtang committed:
Use `@deprecated_args`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @martinwicke for the review. The PR has been updated with `@deprecated_args` used.

drpngx commented:
Any reason to choose `keepdims` instead of `keep_dims`? The latter seems like the better choice to me.

yongtang commented:
@drpngx Not knowing enough history about the code base though I would assume the reason of having `keepdims` instead of `keep_dims` is to align with `numpy`? e.g., 

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.amin.html#numpy.amin

At the moment `keras` backend use `keepdims` while the rest of tensorflow mostly use `keep_dims`.

drpngx commented:
Thank you @yongtang , that makes sense, `keepdims` seems good then.

drpngx commented:
So, sorry to be dense, but does that mean that current code will issue a warning when `tf.reduce_sum(.., keep_dims=True)` will issue a warning but and behave like `tf.reduce_sum(..., keepdims=False)` (the default)?

Don't we want to keep **both** and override `keep_dims` with `keepdims` if both are specified?

yongtang commented:
Thanks @drpngx. The current behavior of this PR is to show a warning in doc string (from the annotation `@deprecated_args`), and accept either `keep_dims` or `keepdims` in the call.

Both `keep_dims` or `keepdims` should work.

However, in case both `keep_dims` and `keepdims` are specified in the same call, then an error will be thrown.

drpngx commented:
Oh, just saw the code. Sounds good.

drpngx commented:
Jenkins, test this please.

yongtang committed:
Update goldens API for compatibility changes

This commit updates goldens API:
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The PR has been updated with 
```
    bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
              --update_goldens True
```
so that `//tensorflow/tools/api/tests:api_compatibility_test` could pass.

martinwicke commented:
Jenkins, test this please.

yongtang commented:
The `Linux CPU Tests Makefile` test might be unrelated I think?

drpngx commented:
this one seems like a download error

Jenkins, test this please.

neilpanchal commented:
Violates PEP guidelines, https://github.com/tensorflow/tensorflow/issues/6815#issuecomment-355110926

I vote for keeping `keep_dims`. There needs to be a strong compelling reason to change. Or support both and get rid of the annoying warnings.
